sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -274,12 +275,30 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
   }
 
   object CartesianProduct extends Strategy {
+    def getSmallSide(left: LogicalPlan, right: LogicalPlan): BuildSide = {
+      if (right.statistics.sizeInBytes < left.statistics.sizeInBytes) {
+        joins.BuildRight
+      } else {
+        joins.BuildLeft
+      }
+    }
+
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+      // If plan can broadcast we use BroadcastNestedLoopJoin, as we know for inner join with true
+      // condition is same as Cartesian.
+      case logical.Join(CanBroadcast(left), right, joinType, condition) =>
+        execution.joins.BroadcastNestedLoopJoin(
+          planLater(left), planLater(right), joins.BuildLeft, joinType, condition) :: Nil
+      case logical.Join(left, CanBroadcast(right), joinType, condition) =>
+        execution.joins.BroadcastNestedLoopJoin(
+          planLater(left), planLater(right), joins.BuildRight, joinType, condition) :: Nil
       case logical.Join(left, right, _, None) =>
-        execution.joins.CartesianProduct(planLater(left), planLater(right)) :: Nil
+        execution.joins.CartesianProduct(planLater(left), planLater(right),
+          getSmallSide(left, right)) :: Nil
       case logical.Join(left, right, Inner, Some(condition)) =>
         execution.Filter(condition,
-          execution.joins.CartesianProduct(planLater(left), planLater(right))) :: Nil
+          execution.joins.CartesianProduct(planLater(left), planLater(right),
+            getSmallSide(left, right))) :: Nil
```

Comment:
Actually I am a little concern about the side switch based on the statistic, as I commented previously. And also as @cloud-fan comment out:
> ```scala
for (x <- rdd1.iterator(currSplit.s1, context);
     y <- rdd2.iterator(currSplit.s2, context)) yield (x, y)
```

What we actually cared is the `average amount of records` in each partition in both sides, and, I don't think we can say, the one take the bigger file size in statistics will also with more `average amount of records` in its partition(most likely the average amount of records in each partition should be same).

Probably we'd better add more statistic info says partition number logical plan or average file size of each partition, and in order not to make confusing for the further improvement, I think we'd better remove this optimization rule for cartesian join. And that's why I didn't do that at #8652

What do you think?